pandas/tests/frame/test_asof.py

```diff
@@ -75,9 +74,34 @@ def test_missing(self):
         result = df.asof('1989-12-31')
 
         expected = Series(index=['A', 'B'], name=Timestamp('1989-12-31'))
-        assert_series_equal(result, expected)
+        tm.assert_series_equal(result, expected)
 
         result = df.asof(to_datetime(['1989-12-31']))
         expected = DataFrame(index=to_datetime(['1989-12-31']),
                              columns=['A', 'B'], dtype='float64')
-        assert_frame_equal(result, expected)
+        tm.assert_frame_equal(result, expected)
+
+    def test_all_nans(self):
+        # GH 15713
+        # DataFrame is all nans
+        result = DataFrame([np.nan]).asof([0])
+        expected = DataFrame([np.nan])
+        tm.assert_frame_equal(result, expected)
+
+        # testing non-default indexes, multiple inputs
+        dates = date_range('1/1/1990', periods=self.N * 3, freq='25s')
+        result = DataFrame(np.nan, index=self.rng, columns=['A']).asof(dates)
+        expected = DataFrame(np.nan, index=dates, columns=['A'])
+        tm.assert_frame_equal(result, expected)
+
+        # testing multiple columns
+        dates = date_range('1/1/1990', periods=self.N * 3, freq='25s')
+        result = DataFrame(np.nan, index=self.rng, columns=['A', 'B', 'C']).asof(dates)
+        expected = DataFrame(np.nan, index=dates, columns=['A', 'B', 'C'])
+        tm.assert_frame_equal(result, expected)
+
+        # testing scalar input
+        date = date_range('1/1/1990', periods=self.N * 3, freq='25s')[0]
+        result = DataFrame(np.nan, index=self.rng, columns=['A']).asof(date)
+        expected = DataFrame(np.nan, index=[date], columns=['A'])
+        tm.assert_frame_equal(result, expected)
```

Comment:
I think a scalar input should result in a Series. That is at least the current behaviour for the working non-NaN case:

```
In [37]: df = pd.DataFrame(np.random.randn(2,2), index=[1,2], columns=['A', 'B'])

In [38]: df
Out[38]: 
          A         B
1 -0.643872  1.375342
2 -0.223192  0.231439

In [39]: df.asof([3])
Out[39]: 
          A         B
3 -0.223192  0.231439

In [40]: df.asof(3)
Out[40]: 
A   -0.223192
B    0.231439
Name: 3, dtype: float64
```